src/awkward/operations/ak_drop_none.py

```diff
@@ -0,0 +1,82 @@
+# BSD 3-Clause License; see https://github.com/scikit-hep/awkward-1.0/blob/main/LICENSE
+
+import awkward as ak
+
+np = ak.nplikes.NumpyMetadata.instance()
+
+
+def drop_none(array, axis=None, highlevel=True, behavior=None):
+    """
+    Args:
+        array: Data in which to remove Nones.
+        axis (None or int): If None, the operation drops Nones at all levels of
+            nesting, returning an array of the same dimension, but without Nones.
+            Otherwise, it drops Nones at a specified depth.
+            The outermost dimension is `0`, followed by `1`, etc.,
+            and negative values count backward from the innermost: `-1` is the
+            innermost dimension, `-2` is the next level up, etc.
+        highlevel (bool): If True, return an #ak.Array; otherwise, return
+            a low-level #ak.contents.Content subclass.
+        behavior (None or dict): Custom #ak.behavior for the output array, if
+            high-level.
+
+    Removes missing values (None) from a given array.
+
+    For example, in the following `array`,
+
+       a = ak.Array([[[0]], [[None]], [[1], None], [[2, None]]])
+
+    The None value will be removed, resulting in
+
+        >>> ak.drop_none(a)
+        <Array [[[0]], [[]], [[1]], [[2]]] type='4 * var * var * int64'>
+
+    The default axis is None, however an axis can be specified:
+
+        >>> ak.drop_none(a,axis=1)
+        <Array [[[0]], [[None]], [[1]], [[2, None]]] type='4 * var * var * ?int64'>
+
+    """
+    with ak._errors.OperationErrorContext(
+        "ak.drop_none",
+        dict(array=array, axis=axis, highlevel=highlevel, behavior=behavior),
+    ):
+        return _impl(array, axis, highlevel, behavior)
+
+
+def _impl(array, axis, highlevel, behavior):
+    layout = ak.operations.to_layout(array, allow_record=False, allow_other=False)
+
+    if layout.is_numpy:
+        return array
+
+    def maybe_drop_none(layout):
+        if layout.is_list:
+            return layout.drop_none()
+        else:
+            return layout
+
+    if axis is None:
+        if layout.is_option:
+            return layout.project()
+
+        def action(layout, continuation, **kwargs):
+            return maybe_drop_none(continuation())
+
+    else:
+
+        def action(layout, depth, depth_context, **kwargs):
+            posaxis = layout.axis_wrap_if_negative(depth_context["posaxis"])
+
+            if posaxis == depth and layout.is_option:
+                return layout.project()
+            elif posaxis == depth and layout.is_list:
+                if layout.content.is_option:
+                    return layout.drop_none()
+
+            depth_context["posaxis"] = posaxis
+
+    depth_context = {"posaxis": axis}
+    out = layout.recursively_apply(action, behavior, depth_context)
```

Comment:
None of the `Content` subclasses have a `recursively_apply` method (without an underscore). What's going on here?

Reply:
I'm wrong: all of the `Content` subclasses have `recursively_apply` (no underscore). I don't know what I was seeing a moment ago.

```python
>>> ak.Array([1, 2, 3]).layout._recursively_apply
<bound method NumpyArray._recursively_apply of <NumpyArray dtype='int64' len='3'>[1 2 3]</NumpyArray>>

>>> ak.Array([{"x": 1}, {"x": 2}, {"x": 3}]).layout.recursively_apply
<bound method Content.recursively_apply of <RecordArray is_tuple='false' len='3'>
    <content index='0' field='x'>
        <NumpyArray dtype='int64' len='3'>[1 2 3]</NumpyArray>
    </content>
</RecordArray>>
```

Reply:
But `Content` does, why is where this is triggered.

Reply:
I see that there's a

```python
return ak._util.wrap(out, behavior, highlevel)
```

and that `highlevel=True` is the default. So why is this happening?

```python
>>> ak.drop_none(ak.Array([1, None, 3]))
<NumpyArray dtype='int64' len='2'>[1 3]</NumpyArray>

>>> ak.drop_none(ak.Array([1, None, 3]), highlevel=True)
<NumpyArray dtype='int64' len='2'>[1 3]</NumpyArray>
```

Reply:
Fixing it. It is because I made an assumption on NumpyArray and returned at the start of the method

Reply:
If there's nothing to drop, it returns a high-level array, regardless of how `highlevel` is set:

```python
>>> ak.drop_none(ak.Array([1, 2, 3]))
<Array [1, 2, 3] type='3 * int64'>
>>> ak.drop_none(ak.Array([1, 2, 3]), highlevel=True)
<Array [1, 2, 3] type='3 * int64'>
>>> ak.drop_none(ak.Array([1, 2, 3]), highlevel=False)
<Array [1, 2, 3] type='3 * int64'>
```